CLID-636: Update AGENTS.md (#1421)
openshift/oc-mirror: AGENTS.md

## Project Overview

> This is the `oc-mirror` repository — an OpenShift client tool for mirroring container registry content for disconnected cluster installs.

Reviewer: ```suggestion
This is the `oc-mirror` repository - an OpenShift client tool (Command Line Interface - CLI) for mirroring container registry content for disconnected cluster installs.
```

Reviewer 2: Maybe just 
```suggestion
This is the `oc-mirror` repository — an OpenShift command line tool for mirroring container registry content for disconnected cluster installs.
```

Reviewer: Adolfo suggestion is simpler, which is better.

### Archive

The archive package handles packaging mirrored images into tar archives (used by MirrorToDisk) and extracting archives back to the filesystem (used by DiskToMirror). This enables the air-gapped workflow where images are transported physically.

### Cluster Resources

After mirroring, oc-mirror generates Kubernetes resources that configure the target cluster to use the mirror registry:

- **IDMS** (ImageDigestMirrorSet) and **ITMS** (ImageTagMirrorSet) — redirect image pulls to the mirror
- **CatalogSource** / **ClusterCatalog** — point the cluster's OLM at mirrored operator catalogs
- **UpdateService** — configure Cincinnati graph for mirrored release channels

### History

The `History` interface (`internal/pkg/history`) tracks which images were synced across runs via `Read()` and `Append()` methods. This enables incremental mirroring — subsequent runs only copy new or updated images.

### Image transport

> The project uses [containers/image](https://github.com/containers/image) for low-level container image transport and copy operations.

Reviewer: ```suggestion
The project uses [podman-container-tools/container-libs](https://github.com/podman-container-tools/container-libs/tree/main/image) for low-level container image transport and copy operations.
```

Author: Thanks @aguidirh , changed

1. **diskToMirror (d2m)** : copy images from tarball to target registry

### Version structure
- **v2** (Current) : code lives in the root directory — **THIS IS WHAT YOU SHOULD WORK ON**
- **v1** (Deprecated) : code lives under the `v1/` folder — **DO NOT MODIFY v1 CODE**

## Project structure

```text
oc-mirror/
├── cmd/oc-mirror/          # CLI entrypoint and embedded data
├── pkg/
│   ├── cli/mirror/         # top-level mirror orchestration (m2m, m2d, d2m)

> │   └── metadata/storage/   # metadata/history persistence

Reviewer: In v2 there is no pkg/cli/mirror and pkg/metadata/storage

Reviewer 2: I am not sure if it's worth to have the file trees here. This might change over time and it will become obsolete. I think agents can easily find out the repository structure dynamically. I think instead we should emphasize the high-level components and architecture (collector, batch worker, etc), their function and why they exist.

wdyt? @aguidirh @dorzel

Reviewer: I like the idea of explaining the high-level packages without having them in a tree structure, in this way when we refactor we don't need to remember to update the tree structure. Moving pieces from one place to another is a very common task and it would make the tree outdated easily.

Author: Makes more sense to do the high level overview, agreed

├── internal/
│   ├── pkg/

> │   │   ├── api/v2alpha1/   # ImageSetConfiguration types and schemas

Reviewer: There are other sub-folders under api, archive, cli, clusterresources, config, history. Also there are some pkgs missing in the list (spinners and version for example, there are others missing). Should we add all of them?

Each folder, even if it is a sub-folder it is considered a different pkg in golang, so maybe should we considere also sub-folders in the structure?

Author: collapsing these into the high level overview we discussed above

│   │   ├── release/        # OpenShift release payload collector
│   │   ├── operator/       # Operator catalog collector (+ filtering)
│   │   ├── helm/           # Helm chart collector
│   │   ├── additional/     # Generic additional image collector
│   │   ├── batch/          # Concurrent image copy worker
│   │   ├── archive/        # Tarball creation/extraction
│   │   ├── mirror/         # Low-level copy interfaces
│   │   ├── image/          # Image reference parsing
│   │   ├── clusterresources/ # IDMS/ITMS/CatalogSource generation
│   │   ├── cincinnati/     # Cincinnati/update graph client
│   │   ├── signature/      # Image signature handling
│   │   ├── delete/         # Image deletion logic
│   │   ├── history/        # Mirror run history tracking
│   │   ├── cli/            # Internal CLI executor
│   │   ├── config/         # Registry auth and config
│   │   └── ...             # log, emoji, spinners, consts, etc.
│   ├── e2e/                # E2E test infrastructure and templates
│   └── testutils/          # Shared test helpers

> ├── tests/                  # Test fixtures (OCI images, configs, caches)

Reviewer: I believe there is a typo here.

```suggestion
├── tests/                  # Test features (OCI images, configs, caches)
```

├── test/e2e/               # E2E test data
├── docs/                   # Documentation (see below)
├── hack/                   # Build and CI scripts
├── images/cli/             # Container image build (Dockerfile support)
└── v1/                     # Deprecated v1 code — DO NOT MODIFY
```

### Documentation tree

```text

> docs/

Reviewer 2: ditto

The `ImageSetConfiguration` types are defined in `internal/pkg/api/v2alpha1/`. See `docs/image-set-examples/image-set-config.yaml` for an example.

## Common development commands

### Building

```bash
make build       # build v1 + v2 binary into ./bin/
make clean       # remove build artifacts
make cross-build # cross-compile for amd64, arm64, ppc64le, s390x
```

**Important**: always use `make build`, not `go build` directly — the Makefile sets required build tags (`json1`, btrfs/libdm exclusions, etc.) and embeds the v1 binary.

Individual cross-build targets are also available: `cross-build-linux-amd64`, `cross-build-linux-arm64`, `cross-build-linux-ppc64le`, `cross-build-linux-s390x`.

### Testing

```bash
make test-unit        # unit tests (./internal/pkg/... -short)
make test-integration # integration tests (runs specific Integration* test funcs)
make cover            # generate HTML coverage report from unit test results
```

Unit tests write coverage to `tests/results/cover.out`. Integration tests write to `tests/results-integration/`.

To run a single test or package directly, you **must** pass the build tags:

```bash

> go test -tags "json1 exclude_graphdriver_devicemapper exclude_graphdriver_btrfs containers_image_openpgp" \

Reviewer: Not so sure if these tags are really needed. Maybe am I missing anything?

Reviewer 3: If someone doesn't have all the deps installed, some of these flags might be needed.

  -short -race -count=1 ./internal/pkg/image/...

go test -tags "json1 exclude_graphdriver_devicemapper exclude_graphdriver_btrfs containers_image_openpgp" \

>   -short -race -count=1 -run TestSpecificName ./internal/pkg/release/...

Reviewer: @adolfo-ab is this the right way to call specific tests when using the new approach with ginko?

Reviewer 2: It's pretty similar, yes

```

### Validation and verification

```bash
make verify  # run golangci-lint
make vet     # run go vet
make format  # check gofmt compliance
make tidy    # run go mod tidy
make sanity  # runs tidy + format + vet, then fails if working tree is dirty
make all     # clean + tidy + build + test (full pipeline)
```

Run `make sanity` before committing — it will fail if there are uncommitted formatting or module changes.

> ## Testing patterns

Reviewer 2: I think we can redirect the agent to docs/testing, which I just added here: https://github.com/openshift/oc-mirror/pull/1422

- **Unit vs integration**: `make test-unit` runs with `-short`. Integration tests check `testing.Short()` and skip themselves, so they only run via `make test-integration`.
- **Mocking**: tests use `github.com/stretchr/testify/mock`. Mock structs are defined alongside the tests that use them (see `internal/pkg/cli/executor_test.go` for examples).

> - **Test fixtures**: stored in `tests/` and referenced via the constant `consts.TestFolder` (`"../../../tests/"`). Fixtures include fake OCI images, caches, configs, and archive data.

Reviewer: typo
```suggestion
- **Test features**: stored in `tests/` and referenced via the constant `consts.TestFolder` (`"../../../tests/"`). Features include fake OCI images, caches, configs, and archive data.
```
